### ISSUE TITLE
Nullable attributes don't modify indexer parameter initial nullable state

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#73484]
**Version Used**: fca6e1fcdcded85e69cc32e15acfb6820cd45597

**Steps to Reproduce**:

```cs
[Fact]
public void AllowNull_Indexer()
{
    var source = """
        #nullable enable

        using System;
        using System.Diagnostics.CodeAnalysis;

        class C
        {
            public static void Main()
            {
                var c = new C();
                try
                {
                    _ = c[null]; // no warning
                }
                catch
                {
                    Console.Write(1);
                }
            }

            public string this[[AllowNull] string s]
            {
                get
                {
                    return s.ToString(); // expected a warning
                }
            }
        }
        """;

    var verifier = CompileAndVerify([source, AllowNullAttributeDefinition], expectedOutput: "1");
    verifier.VerifyDiagnostics();
}
```

**Expected Behavior**: A warning is reported on the indicated line

**Actual Behavior**: No warning is reported in the sample
